src/kagglehub/gcs_upload.py

```diff
@@ -138,51 +148,111 @@ def _upload_blob(file_path: str, model_type: str) -> str:
-    return size
-
-
-def zip_files(source_path_obj: Path, zip_path: Path) -> List[int]:
-    files = [file for file in source_path_obj.rglob("*") if file.is_file()]
-    args = [(file, zip_path, source_path_obj) for file in files]
-
-    with Pool() as pool:
-        sizes = pool.map(zip_file, args)
-    return sizes
-
-
-def upload_files(source_path: str, model_type: str) -> List[str]:
-    source_path_obj = Path(source_path)
-    with TemporaryDirectory() as temp_dir:
-        temp_dir_path = Path(temp_dir)
-        total_size = 0
-
-        if source_path_obj.is_dir():
-            for file_path in source_path_obj.rglob("*"):
-                if file_path.is_file():
-                    total_size += file_path.stat().st_size
-        elif source_path_obj.is_file():
-            total_size = source_path_obj.stat().st_size
-        else:
-            path_error_message = "The source path does not point to a valid file or directory."
-            raise ValueError(path_error_message)
-
-        with tqdm(total=total_size, desc="Zipping", unit="B", unit_scale=True, unit_divisor=1024) as pbar:
-            if source_path_obj.is_dir():
-                zip_path = temp_dir_path / "archive.zip"
-                sizes = zip_files(source_path_obj, zip_path)
-                for size in sizes:
-                    pbar.update(size)
-                upload_path = str(zip_path)
-            elif source_path_obj.is_file():
-                temp_file_path = temp_dir_path / source_path_obj.name
-                shutil.copy(source_path_obj, temp_file_path)
-                pbar.update(temp_file_path.stat().st_size)
-                upload_path = str(temp_file_path)
-
-        return [token for token in [_upload_blob(upload_path, model_type)] if token]
+def upload_files(folder: str, model_type: str, quiet: bool = False) -> List[str]:  # noqa: FBT002, FBT001
+    # Count the total number of files
+    file_count = 0
+    for _, _, files in os.walk(folder):
+        file_count += len(files)
+
+    if file_count > MAX_FILES_TO_UPLOAD:
+        if not quiet:
+            logger.info(f"More than {MAX_FILES_TO_UPLOAD} files detected, creating a zip archive...")
+
+        with TemporaryDirectory() as temp_dir:
+            zip_path = os.path.join(temp_dir, TEMP_ARCHIVE_FILE)
+            with zipfile.ZipFile(zip_path, "w") as zipf:
+                for root, _, files in os.walk(folder):
+                    for file in files:
+                        file_path = os.path.join(root, file)
+                        zipf.write(file_path, os.path.relpath(file_path, folder))
+
+            # Upload the zip file
+            return {
+                "files": [
+                    token
+                    for token in [_upload_file_or_folder(temp_dir, TEMP_ARCHIVE_FILE, model_type, quiet)]
+                    if token is not None
+                ],
+                "directories": [],
+            }
+
+    root_dict = {'files': [], 'directories': []}
```

Comment:
I would instantiate a `UploadDirectoryInfo` type for the root directory. Add a constructor to `UploadDirectoryInfo` to initialize the files & directories array to empty arrays.